Mathlib/Tactic/Linter/HaveLetLinter.lean

```diff
@@ -0,0 +1,126 @@
+            * 1 -- active only on noisy declarations;\n\
+            * 2 or more -- always active."
+}
+
+namespace haveLet
+
+/-- find the `have` syntax. -/
+partial
+def isHave? : Syntax → Bool
+  | .node _ ``Lean.Parser.Tactic.tacticHave_ _ => true
+  |_ => false
+
+end haveLet
+
+end Mathlib.Linter
+
+namespace Mathlib.Linter.haveLet
+
+/-- a monadic version of `Lean.Elab.InfoTree.foldInfo`.
+Used to infer types inside a `CommandElabM`. -/
+def InfoTree.foldInfoM {α m} [Monad m] (f : ContextInfo → Info → α → m α) (init : α) :
+    InfoTree → m α :=
+  InfoTree.foldInfo (fun ctx i ma => do f ctx i (← ma)) (pure init)
+
+/-- given a `ContextInfo`, a `LocalContext` and an `Array` of `Expr`essions `es`,
+`areProp_toFormat` creates a `MetaM` context, and returns an array of pairs consisting of
+* a `Bool`ean, answering the question of whether the Type of `e` is a `Prop` or not, and
+* the pretty-printed `Format` of `e`
+for each `Expr`ession `e` in `es`.
+Concretely, `areProp_toFormat` runs `inferType` in `CommandElabM`.
+This is the kind of monadic lift that `nonPropHaves` uses to decide whether the Type of a `have`
+is in `Prop` or not.
+The output `Format` is just so that the linter displays a better message. -/
+def areProp_toFormat (ctx : ContextInfo) (lc : LocalContext) (es : Array Expr) :
+    CommandElabM (Array (Bool × Format)) := do
+  ctx.runMetaM lc do
+    es.mapM fun e => do
+      let typ ← inferType (← instantiateMVars e)
+      return (typ.isProp, ← ppExpr e)
+
+/-- returns the `have` syntax whose corresponding hypothesis does not have Type `Prop` and
+also a `Format`ted version of the corresponding Type. -/
+partial
+def nonPropHaves : InfoTree → CommandElabM (Array (Syntax × Format)) :=
+  InfoTree.foldInfoM (init := #[]) fun ctx info args => return args ++ (← do
+    let .ofTacticInfo i := info | return #[]
+    let stx := i.stx
+    let .original .. := stx.getHeadInfo | return #[]
+    unless isHave? stx do return #[]
+    let mctx := i.mctxAfter
+    let mvdecls := (i.goalsAfter.map (mctx.decls.find? ·)).reduceOption
+    let _ : Ord MetavarDecl := { compare := (compare ·.index ·.index) }
+    -- we extract the `MetavarDecl` with largest index after a `have`, since this one
+    -- holds information about the metavariable where `have` introduces the new hypothesis.
+    let largestIdx := mvdecls.toArray.qsort (·.index > ·.index)
+    -- the relevant `LocalContext`
+    let lc := (largestIdx.getD 0 default).lctx
+    -- we also accumulate all `fvarId`s from all local contexts before the use of `have`
+    -- so that we can then isolate the `fvarId`s that are created by `have`
+    let mvdeclsOld := (i.goalsBefore.map (mctx.decls.find? ·)).reduceOption
+    let lctxOld := mvdeclsOld.map (·.lctx)
+    let declsOld := (lctxOld.map (·.decls.toList.reduceOption)).join
+    let fvAssOld := declsOld.map (·.fvarId)
+    let oldDecls := lc.decls.toList.reduceOption.filter (! fvAssOld.contains ·.fvarId)
```

Comment:
`declsOld` and `oldDecls` is a bit unfortunate naming. Am I understanding this correct that `oldDecls` are all `LocalDecl`s that are introduced newly by the tactic in question? If so, wouldn't `declsNew` be more fitting?